File: frontend/src/lib/xray/subscription-profile.ts
import type { StreamSettings } from '@/schemas/api/inbound';
import type { ExternalProxyEntry } from '@/schemas/protocols/stream/external-proxy';


export function createSubscriptionProfileDraft(
  defaultAddress = '',
  defaultPort = 443,
): ExternalProxyEntry {
  const safePort = Number.isInteger(defaultPort) && defaultPort > 0 && defaultPort <= 65535
    ? defaultPort
    : 443;

  return {
    enabled: true,
    remark: '',
    dest: defaultAddress,
    port: safePort,
    network: 'same',
    security: 'same',
    forceTls: 'same',
    excludeFromSubTypes: [],
    mihomoX25519: false,
    shuffleHost: false,
  };
}

export interface SubscriptionProfileEndpoint {
  address: string;
  port: number;
  remark: string;
  streamSettings: StreamSettings;
  profile: ExternalProxyEntry | null;
}

const TRANSPORT_KEYS = [
  'tcpSettings',
  'kcpSettings',
  'wsSettings',
  'grpcSettings',
  'httpupgradeSettings',
  'xhttpSettings',
  'hysteriaSettings',
] as const;

type MutableStream = Record<string, unknown>;
type MutableProfile = ExternalProxyEntry & Record<string, unknown>;

function jsonClone<T>(value: T): T {
  return JSON.parse(JSON.stringify(value)) as T;
}

function transportKey(network: string): (typeof TRANSPORT_KEYS)[number] | null {
  switch (network) {
    case 'tcp': return 'tcpSettings';
    case 'kcp': return 'kcpSettings';
    case 'ws': return 'wsSettings';
    case 'grpc': return 'grpcSettings';
    case 'httpupgrade': return 'httpupgradeSettings';
    case 'xhttp': return 'xhttpSettings';
    case 'hysteria': return 'hysteriaSettings';
    default: return null;
  }
}

function defaultTransportSettings(network: string): Record<string, unknown> {
  switch (network) {
    case 'tcp':
      return { acceptProxyProtocol: false, header: { type: 'none' } };
    case 'ws':
      return { acceptProxyProtocol: false, path: '/', host: '', headers: {}, heartbeatPeriod: 0 };
    case 'grpc':
      return { serviceName: '', authority: '', multiMode: false };
    case 'httpupgrade':
      return { acceptProxyProtocol: false, path: '/', host: '', headers: {} };
    case 'xhttp':
      return { path: '/', host: '', mode: 'auto', headers: {} };
    default:
      return {};
  }
}

function defaultTlsSettings(): Record<string, unknown> {
  return {
    serverName: '',
    alpn: [],
    settings: {
      fingerprint: 'chrome',
      echConfigList: '',
      pinnedPeerCertSha256: [],
      allowInsecure: false,
    },
  };
}

function defaultRealitySettings(): Record<string, unknown> {
  return {
    serverNames: [],
    shortIds: [],
    settings: {
      publicKey: '',
      fingerprint: 'chrome',
      serverName: '',
      spiderX: '/',
      mldsa65Verify: '',
    },
  };
}

function applyLegacyTlsFields(profile: MutableProfile, stream: MutableStream): void {
  const tls = (stream.tlsSettings && typeof stream.tlsSettings === 'object')
    ? stream.tlsSettings as Record<string, unknown>
    : defaultTlsSettings();
  stream.tlsSettings = tls;

  if (profile.sni) tls.serverName = profile.sni;
  if (Array.isArray(profile.alpn) && profile.alpn.length > 0) tls.alpn = [...profile.alpn];

  const settings = (tls.settings && typeof tls.settings === 'object')
    ? tls.settings as Record<string, unknown>
    : {};
  tls.settings = settings;
  if (profile.fingerprint) settings.fingerprint = profile.fingerprint;
  if (Array.isArray(profile.pinnedPeerCertSha256) && profile.pinnedPeerCertSha256.length > 0) {
    settings.pinnedPeerCertSha256 = [...profile.pinnedPeerCertSha256];
  }
  if (profile.echConfigList) settings.echConfigList = profile.echConfigList;
  if (typeof profile.allowInsecure === 'boolean') settings.allowInsecure = profile.allowInsecure;
}

export function effectiveSubscriptionProfileStream(
  base: StreamSettings,
  profile: ExternalProxyEntry,
): StreamSettings {
  const stream = jsonClone(base) as unknown as MutableStream;
  delete stream.externalProxy;
  const mutableProfile = profile as MutableProfile;

  const baseNetwork = typeof stream.network === 'string' ? stream.network : '';
  const selectedNetwork = profile.network && profile.network !== 'same'
    ? profile.network
    : baseNetwork;
  const selectedTransportKey = transportKey(selectedNetwork);
  if (selectedTransportKey) {
    if (selectedNetwork !== baseNetwork) {
      for (const key of TRANSPORT_KEYS) delete stream[key];
      stream.network = selectedNetwork;
      stream[selectedTransportKey] = defaultTransportSettings(selectedNetwork);
    }
    const profileSettings = mutableProfile[selectedTransportKey];
    if (profileSettings && typeof profileSettings === 'object') {
      stream[selectedTransportKey] = jsonClone(profileSettings);
    } else if (!stream[selectedTransportKey]) {
      stream[selectedTransportKey] = defaultTransportSettings(selectedNetwork);
    }
  }

  let security = profile.security;
  if (!security || security === 'same') security = profile.forceTls;
  if (!security || security === 'same') {
    security = typeof stream.security === 'string' ? stream.security as 'none' | 'tls' | 'reality' : 'none';
  }

  switch (security) {
    case 'none':
      stream.security = 'none';
      delete stream.tlsSettings;
      delete stream.realitySettings;
      break;
    case 'tls':
      stream.security = 'tls';
      delete stream.realitySettings;
      if (profile.tlsSettings) stream.tlsSettings = jsonClone(profile.tlsSettings);
      else if (!stream.tlsSettings) stream.tlsSettings = defaultTlsSettings();
      applyLegacyTlsFields(mutableProfile, stream);
      break;
    case 'reality':
      stream.security = 'reality';
      delete stream.tlsSettings;
      if (profile.realitySettings) stream.realitySettings = jsonClone(profile.realitySettings);
      else if (!stream.realitySettings) stream.realitySettings = defaultRealitySettings();
      break;
    default:
      break;
  }

  if (profile.finalmask) stream.finalmask = jsonClone(profile.finalmask);
  return stream as unknown as StreamSettings;
}

export function expandSubscriptionProfileEndpoints(
  streamSettings: StreamSettings,
  defaultAddress: string,
  defaultPort: number,
): SubscriptionProfileEndpoint[] {
  const profiles = streamSettings.externalProxy;
  if (!profiles || profiles.length === 0) {
    const stream = jsonClone(streamSettings) as unknown as MutableStream;
    delete stream.externalProxy;
    return [{
      address: defaultAddress,
      port: defaultPort,
      remark: '',
      streamSettings: stream as unknown as StreamSettings,
      profile: null,
    }];
  }

  return profiles
    .filter((profile) => profile.enabled !== false)
    .map((profile) => ({
      address: profile.dest?.trim() || defaultAddress,
      port: Number.isInteger(profile.port) && profile.port > 0 && profile.port <= 65535
        ? profile.port
        : defaultPort,
      remark: profile.remark ?? '',
      streamSettings: effectiveSubscriptionProfileStream(streamSettings, profile),
      profile,
    }));
}

export interface DefaultSubscriptionPortPair {
  inboundPort: number | null;
  profilePort: number | null;
}

export interface DefaultSubscriptionPortSyncState
  extends DefaultSubscriptionPortPair {
  linked: boolean;
  pending: DefaultSubscriptionPortPair | null;
}

export interface DefaultSubscriptionPortSyncPlan {
  state: DefaultSubscriptionPortSyncState;
  setInboundPort?: number;
  setProfilePort?: number;
}

export function normalizeSubscriptionPort(value: unknown): number | null {
  if (
    typeof value !== 'number'
    || !Number.isInteger(value)
    || value < 1
    || value > 65535
  ) {
    return null;
  }
  return value;
}

function portPairsEqual(
  left: DefaultSubscriptionPortPair,
  right: DefaultSubscriptionPortPair,
): boolean {
  return left.inboundPort === right.inboundPort
    && left.profilePort === right.profilePort;
}

function portsAreLinked(pair: DefaultSubscriptionPortPair): boolean {
  return pair.inboundPort !== null
    && pair.profilePort !== null
    && pair.inboundPort === pair.profilePort;
}

function createPortSyncState(
  pair: DefaultSubscriptionPortPair,
): DefaultSubscriptionPortSyncState {
  return {
    ...pair,
    linked: portsAreLinked(pair),
    pending: null,
  };
}

// Profile zero is the default endpoint. While its port equals the inbound
// port, editing either side keeps both synchronized. Existing imported or
// intentionally divergent profiles remain independent until equalized again.
export function planDefaultSubscriptionPortSync(
  previous: DefaultSubscriptionPortSyncState | null,
  current: DefaultSubscriptionPortPair,
): DefaultSubscriptionPortSyncPlan {
  if (previous === null) {
    return { state: createPortSyncState(current) };
  }

  if (previous.pending !== null) {
    if (portPairsEqual(current, previous.pending)) {
      return {
        state: {
          ...current,
          linked: true,
          pending: null,
        },
      };
    }

    // setFieldValue can produce an intermediate render. Do not reverse the
    // synchronization while waiting for the target field to update.
    return { state: previous };
  }

  const inboundChanged = current.inboundPort !== previous.inboundPort;
  const profileChanged = current.profilePort !== previous.profilePort;

  if (!previous.linked) {
    return { state: createPortSyncState(current) };
  }

  if (!inboundChanged && !profileChanged) {
    return { state: previous };
  }

  if (inboundChanged && !profileChanged) {
    if (current.inboundPort === null) {
      return { state: previous };
    }

    const target = {
      inboundPort: current.inboundPort,
      profilePort: current.inboundPort,
    };

    return {
      state: {
        ...target,
        linked: true,
        pending: target,
      },
      setProfilePort: current.inboundPort,
    };
  }

  if (profileChanged && !inboundChanged) {
    if (current.profilePort === null) {
      return { state: previous };
    }

    const target = {
      inboundPort: current.profilePort,
      profilePort: current.profilePort,
    };

    return {
      state: {
        ...target,
        linked: true,
        pending: target,
      },
      setInboundPort: current.profilePort,
    };
  }

  // Initialization/import may change both values simultaneously.
  return { state: createPortSyncState(current) };
}

File: internal/sub/subscription_profile.go
package sub

import (
	"encoding/json"
	"math"
	"strings"
)

// subscriptionEndpoint is the client-facing endpoint produced from one
// streamSettings.externalProxy entry. The name is kept for wire compatibility,
// but SECX treats each entry as a complete subscription profile: address/port
// plus optional transport, security and client-side stream overrides.
type subscriptionEndpoint struct {
	Address        string
	Port           int
	Remark         string
	Stream         map[string]any
	MuxOverride    any
	HasMuxOverride bool
}

var subscriptionTransportKeys = []string{
	"tcpSettings",
	"kcpSettings",
	"wsSettings",
	"grpcSettings",
	"httpupgradeSettings",
	"xhttpSettings",
	"hysteriaSettings",
}

// expandSubscriptionEndpoints returns exactly one default endpoint when no
// profile list is configured. Once a profile list exists, only enabled entries
// are returned. Therefore an explicitly configured list with every profile
// disabled intentionally contributes zero subscription configurations.
func expandSubscriptionEndpoints(baseStream map[string]any, defaultAddress string, defaultPort int) []subscriptionEndpoint {
	raw, exists := baseStream["externalProxy"]
	profiles, ok := raw.([]any)
	if !exists || !ok || len(profiles) == 0 {
		return []subscriptionEndpoint{{
			Address: defaultAddress,
			Port:    defaultPort,
			Stream:  cloneJSONMapWithoutExternalProxy(baseStream),
		}}
	}

	out := make([]subscriptionEndpoint, 0, len(profiles))
	for _, rawProfile := range profiles {
		profile, ok := rawProfile.(map[string]any)
		if !ok || profile == nil {
			continue
		}
		if enabled, present := profile["enabled"].(bool); present && !enabled {
			continue
		}

		address := strings.TrimSpace(stringValue(profile["dest"]))
		if address == "" {
			address = defaultAddress
		}
		port := intValue(profile["port"])
		if port < 1 || port > 65535 {
			port = defaultPort
		}

		endpoint := subscriptionEndpoint{
			Address: address,
			Port:    port,
			Remark:  stringValue(profile["remark"]),
			Stream:  effectiveSubscriptionProfileStream(baseStream, profile),
		}
		if mux, exists := profile["mux"]; exists {
			endpoint.HasMuxOverride = true
			endpoint.MuxOverride = deepCloneJSON(mux)
		}
		out = append(out, endpoint)
	}
	return out
}

func effectiveSubscriptionProfileStream(baseStream, profile map[string]any) map[string]any {
	stream := cloneJSONMapWithoutExternalProxy(baseStream)

	applyProfileTransport(stream, profile)
	applyProfileSecurity(stream, profile)

	if finalMask, ok := profile["finalmask"].(map[string]any); ok {
		stream["finalmask"] = deepCloneJSON(finalMask)
	}
	return stream
}

func applyProfileTransport(stream, profile map[string]any) {
	baseNetwork := stringValue(stream["network"])
	profileNetwork := strings.ToLower(strings.TrimSpace(stringValue(profile["network"])))
	if profileNetwork == "" || profileNetwork == "same" {
		profileNetwork = baseNetwork
	}
	if profileNetwork == "" {
		return
	}

	settingsKey := transportSettingsKey(profileNetwork)
	if settingsKey == "" {
		return
	}

	if profileNetwork != baseNetwork {
		for _, key := range subscriptionTransportKeys {
			delete(stream, key)
		}
		stream["network"] = profileNetwork
		stream[settingsKey] = defaultTransportSettings(profileNetwork)
	}

	if settings, ok := profile[settingsKey].(map[string]any); ok {
		stream[settingsKey] = deepCloneJSON(settings)
	} else if _, ok := stream[settingsKey].(map[string]any); !ok {
		stream[settingsKey] = defaultTransportSettings(profileNetwork)
	}
}

func applyProfileSecurity(stream, profile map[string]any) {
	security := strings.ToLower(strings.TrimSpace(stringValue(profile["security"])))
	if security == "" || security == "same" {
		// Backward compatibility with the original externalProxy.forceTls field.
		security = strings.ToLower(strings.TrimSpace(stringValue(profile["forceTls"])))
	}
	if security == "" || security == "same" {
		security = strings.ToLower(strings.TrimSpace(stringValue(stream["security"])))
	}
	if security == "" {
		security = "none"
	}

	switch security {
	case "none":
		stream["security"] = "none"
		delete(stream, "tlsSettings")
		delete(stream, "realitySettings")
	case "tls":
		stream["security"] = "tls"
		delete(stream, "realitySettings")
		if settings, ok := profile["tlsSettings"].(map[string]any); ok {
			stream["tlsSettings"] = deepCloneJSON(settings)
		} else if _, ok := stream["tlsSettings"].(map[string]any); !ok {
			stream["tlsSettings"] = defaultProfileTLSSettings()
		}
		applyLegacyExternalProxyTLSFields(profile, stream)
	case "reality":
		stream["security"] = "reality"
		delete(stream, "tlsSettings")
		if settings, ok := profile["realitySettings"].(map[string]any); ok {
			stream["realitySettings"] = deepCloneJSON(settings)
		} else if _, ok := stream["realitySettings"].(map[string]any); !ok {
			stream["realitySettings"] = defaultProfileRealitySettings()
		}
	default:
		// Unknown profile values must never corrupt the parent stream.
		return
	}
}

func applyLegacyExternalProxyTLSFields(profile, stream map[string]any) {
	tlsSettings, _ := stream["tlsSettings"].(map[string]any)
	if tlsSettings == nil {
		tlsSettings = defaultProfileTLSSettings()
		stream["tlsSettings"] = tlsSettings
	}

	if sni := strings.TrimSpace(stringValue(profile["sni"])); sni != "" {
		tlsSettings["serverName"] = sni
	}
	if alpn, ok := stringSliceValue(profile["alpn"]); ok {
		tlsSettings["alpn"] = alpn
	}

	clientSettings, _ := tlsSettings["settings"].(map[string]any)
	if clientSettings == nil {
		clientSettings = map[string]any{}
		tlsSettings["settings"] = clientSettings
	}
	if fp := strings.TrimSpace(stringValue(profile["fingerprint"])); fp != "" {
		clientSettings["fingerprint"] = fp
	}
	if pins, ok := stringSliceValue(profile["pinnedPeerCertSha256"]); ok {
		clientSettings["pinnedPeerCertSha256"] = pins
	}
	if ech := strings.TrimSpace(stringValue(profile["echConfigList"])); ech != "" {
		clientSettings["echConfigList"] = ech
	}
	if allowInsecure, ok := profile["allowInsecure"].(bool); ok {
		clientSettings["allowInsecure"] = allowInsecure
	}
}

func cloneJSONMapWithoutExternalProxy(src map[string]any) map[string]any {
	cloned, _ := deepCloneJSON(src).(map[string]any)
	if cloned == nil {
		cloned = map[string]any{}
	}
	delete(cloned, "externalProxy")
	return cloned
}

func deepCloneJSON(value any) any {
	data, err := json.Marshal(value)
	if err != nil {
		return value
	}
	var out any
	if err := json.Unmarshal(data, &out); err != nil {
		return value
	}
	return out
}

func transportSettingsKey(network string) string {
	switch network {
	case "tcp":
		return "tcpSettings"
	case "kcp":
		return "kcpSettings"
	case "ws":
		return "wsSettings"
	case "grpc":
		return "grpcSettings"
	case "httpupgrade":
		return "httpupgradeSettings"
	case "xhttp":
		return "xhttpSettings"
	case "hysteria":
		return "hysteriaSettings"
	default:
		return ""
	}
}

func defaultTransportSettings(network string) map[string]any {
	switch network {
	case "tcp":
		return map[string]any{"acceptProxyProtocol": false, "header": map[string]any{"type": "none"}}
	case "ws":
		return map[string]any{"acceptProxyProtocol": false, "path": "/", "host": "", "headers": map[string]any{}, "heartbeatPeriod": float64(0)}
	case "grpc":
		return map[string]any{"serviceName": "", "authority": "", "multiMode": false}
	case "httpupgrade":
		return map[string]any{"acceptProxyProtocol": false, "path": "/", "host": "", "headers": map[string]any{}}
	case "xhttp":
		return map[string]any{"path": "/", "host": "", "mode": "auto", "headers": map[string]any{}}
	case "kcp":
		return map[string]any{}
	case "hysteria":
		return map[string]any{}
	default:
		return map[string]any{}
	}
}

func defaultProfileTLSSettings() map[string]any {
	return map[string]any{
		"serverName": "",
		"alpn":       []any{},
		"settings": map[string]any{
			"fingerprint":          "chrome",
			"echConfigList":        "",
			"pinnedPeerCertSha256": []any{},
			"allowInsecure":        false,
		},
	}
}

func defaultProfileRealitySettings() map[string]any {
	return map[string]any{
		"serverNames": []any{},
		"shortIds":    []any{},
		"settings": map[string]any{
			"publicKey":     "",
			"fingerprint":   "chrome",
			"serverName":    "",
			"spiderX":       "/",
			"mldsa65Verify": "",
		},
	}
}

func stringValue(value any) string {
	valueString, _ := value.(string)
	return valueString
}

func intValue(value any) int {
	switch v := value.(type) {
	case int:
		return v
	case int64:
		return int(v)
	case float64:
		if math.IsNaN(v) || math.IsInf(v, 0) {
			return 0
		}
		return int(v)
	case json.Number:
		i, _ := v.Int64()
		return int(i)
	default:
		return 0
	}
}

func stringSliceValue(value any) ([]any, bool) {
	switch values := value.(type) {
	case []any:
		out := make([]any, 0, len(values))
		for _, item := range values {
			if itemString, ok := item.(string); ok && strings.TrimSpace(itemString) != "" {
				out = append(out, itemString)
			}
		}
		return out, len(out) > 0
	case []string:
		out := make([]any, 0, len(values))
		for _, item := range values {
			if strings.TrimSpace(item) != "" {
				out = append(out, item)
			}
		}
		return out, len(out) > 0
	default:
		return nil, false
	}
}
